### PR TITLE
Reduce universe redeclarations (catching AlreadyDeclared)

### PR DESCRIPTION
**Pull request by SkySkimmer**
NB: Include is still causing repeat declarations.

### CODE REVIEW

File: kernel/term_typing.ml [hunk 1]
```diff
@@ -267,6 +267,8 @@ let infer_declaration (type a) ~(trust : a trust) env (dcl : a constant_entry) =
       let tyj = infer_type env typ in
       let proofterm =
         Future.chain body (fun ((body,uctx),side_eff) ->
+          (* don't redeclare universes which are declared for the type *)
+          let uctx = Univ.ContextSet.diff uctx univs in
```
Comment (SkySkimmer) [inline kernel/term_typing.ml]: Not sure if this is the right place to do this, but it seemed cleaner than doing it in the proof_global function and it makes OK sense to allow the redundancy in the entry.

**Comment by SkySkimmer** [comment]
BTW allowing redeclarations for monomorphic definitions is fine but for polymorphic definitions it seems inconsistent, allowing to capture global universes.
ie a plugin could produce something like
~~~coq
Universe i.
Definition Ti := Type@{i}.
Polymorphic Definition T@{i (* this is the same as the global i *)|(*no constraints*)} : Ti -> Type@{i} := fun x => x.
Check T@{j}. (* T@{j} : Ti -> Type@{j} (* j |= *) *)
(* !!! the substitution of j doesn't go through Ti *)
~~~
I haven't tried to make such a plugin so maybe something prevents it, but it's definitely dangerous.

After this PR it would produce an anomaly. However the checker may still allow it (something for a future PR) (to trick the checker you may need to use `Var 0` as your global universe since the checker sees universe-abstracted terms).

**Comment by SkySkimmer** [comment]
Inconsistency from plugin confirmed, see test.

**Comment by SkySkimmer** [comment]
Not sure if we want to actually include the test, testing for inconsistencies through plugins is a bit complex. And testing inconsistencies in the checker is essentially impossible.

**Comment by ejgallego** [comment]
> Not sure if we want to actually include the test, testing for inconsistencies through plugins is a bit complex.

Why?

**Review by JasonGross — DISMISSED** [review]
This merits a line in dev/doc/critical-bugs, right?

**Comment by SkySkimmer** [comment]
Only 12204ee and 2385b5c are bug related, the others are just cleanups.

**Review by Zimmi48 — APPROVED** [review]
Approving the doc part.

**Review by mattam82 — APPROVED** [review]
The fix looks good to me. I just think we should go further and disallow Level.Var to appear in monomorphic constant or inductive contexts altogether. @SkySkimmer is Include the only remaining case where redeclarations happen now?

**Comment by SkySkimmer** [comment]
>The fix looks good to me. I just think we should go further and disallow Level.Var to appear in monomorphic constant or inductive contexts altogether. 

Next PR I guess.

>@SkySkimmer is Include the only remaining case where redeclarations happen now?

I don't know.

**Comment by Zimmi48** [comment]
@SkySkimmer Shouldn't this come with a CHANGES entry? If yes, add it to the 8.8.2 section in both this PR and #8360.

**Comment by SkySkimmer** [comment]
Added changes

**Comment by Zimmi48** [comment]
Great thanks! @maximedenes you're the only code owner whose review is missing from this (critical) PR.

**Review by maximedenes — COMMENTED** [review]
I declare myself incompetent.

**Comment by Zimmi48** [comment]
@mattam82 ping

**Comment by silene** [comment]
If I am not mistaken, this is already part of 8.9, so please complain if I am wrongly removing from the project.

**Pull request merged**